### PR TITLE
Enable clients to name their informers

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR expands the interface to informers so that a client creating one can give it a name.

#### Which issue(s) this PR fixes:

This partially addresses #111790

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
I could not see a way to enable a client of SharedInformerFactory to control the informer name before the construction of the informer without changing the relevant code generator and a lot of generated code.  So I opted for the easy way out, adding a way to change an informer's name.  I do not love this, but suspect it is the least bad choice.

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 1]
```diff
@@ -133,6 +133,10 @@ import (
 // state, except that its ResourceVersion is replaced with a
 // ResourceVersion in which the object is actually absent.
 type SharedInformer interface {
+	// SetName can be called before Run to modify the name of the informer.
+	// Returns true on success, false if called after Run.
+	SetName(string) bool
+
```
Comment (alexzielenski) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go:139]: Is the ability to mutate name after construction necessary for your usage case? My intuition would be that the field added to `SharedIndexInformerOptions` was sufficient for naming.

Comment (alexzielenski) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go:139]: just saw this https://github.com/kubernetes/kubernetes/pull/120578#issuecomment-1714591855

Comment (alexzielenski) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go:139]: @MikeSpreitzer can you help me understand your use case a bit more? I wouldve assume the `SharedInformerFactory` would name all of its informers by the name of the resource, possibly with a prefix identifying the factory itself. Is that sufficient for you

Comment (MikeSpreitzer) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go:139]: Do you mean the `SharedInformerOption` at https://github.com/kubernetes/client-go/blob/v0.29.0/informers/factory.go#L54 ? That is in generated code, so changing it involves changing the generator.

That approach would also require work on https://github.com/kubernetes/client-go/blob/v0.29.0/informers/generic.go#L98

The use cases are mentioned at the start of #111790

Comment (MikeSpreitzer) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go:139]: See #122654 for the approach that involves changing the informer generator.

**Comment by MikeSpreitzer** [comment]
@alexzielenski : can you consult with the reviewers of #122654 and agree on which approach to pursue?

**Pull request closed without merging**